src/python/pants/backend/adhoc/target_types.py

```diff
@@ -253,6 +254,33 @@ class AdhocToolOutputRootDirField(StringField):
     )
 
 
+class AdhocToolCacheScopeField(StringField):
```

Comment:
Introduction of cache scope should be moved to a separate PR if we decide to keep it. It was necessary to provide a way for tests to set `cache_scope` to `ProcessCacheScope.PER_SESSION` in order to avoid cached results from prior tests runs from being used for a current test run.

This points to a more fundamental problem: Will workspace executions which modify the workspace be inherently a problem if the user is looking for them to always execute?

Reply:
Maybe dumb question, but why isn't the caching strategy determined by the environment? I.e., local is cacheable and workspace is not? Do we need this extra degree of freedom?

Reply:
If a user does have a reproducible process invoked in the workspace environment (e.g., Bazel), shouldn't Pants cache the output then?

Reply:
There is two sort of issues here:
1. Should processes invoked in workspace environments be cacheable? My thought is yes they should. (Invalidations may occur as a result of the metadata-based invalidation to come from https://github.com/pantsbuild/pants/pull/20914.)
2. The tests of the workspace environment fail when they invoke a non-idempotent process which writes a file into the workspace and which the test checks for to ensure execution actually happened in the workspace). Running that non-idempotent test multipe times fails unless the cache scope is "per session." Maybe this is a code smell in the test?

Reply:
What key would we cache against though? 

I can see not rerunning the process if the invalidation globs haven't changed from the last time we ran the process, but that is not the same as caching. In other words, of we run the process in state A, then as long as we remain in state A we don't rerun it. Then we switch to state B, so obviously we must rerun the process (and we do because the file mtimes have changed). Then we switch back to state A. Normally we'd retrieve the result from cache, but for workspace environments I think we have to rerun the process, no? All we know is that mtimes have changed, we don't fingerprint anything, so there's nothing to compute a cache key out of. 

Or am I fundamentally misunderstanding the proposition here?

Reply:
This PR does not introduce the metadata-based cache key. Thus, you may want to review https://github.com/pantsbuild/pants/pull/20914 which does introduce the cache key: namely a hash of the "full stats" of the files on which the targets are depending via a `invalidation_globs` field. This is injected into the applicable `Process` as an environment variable.

Reply:
Yeah, I've looked at #20914, and unless I'm misreading it, a FullPathStat is not able to generate a useful cache key, because caching against an mtime is not useful - we will never see the same mtime again after changing it once. And if we omit the mtime from the cache key then we have no knowledge of a change occurring.

I'm distinguishing invalidation, which we can do via mtime, vs actual process result caching, which we cannot.

I guess we can trivially implement invalidation via the cache, is that what we're actually talking about here?

Reply:
My main concern is avoiding Pants always invoking the third party tool in order for the tool to confirm that no changes have been made. This goal may be more focused on the pantsd memoization cache than the regular execution cache.

`mtime` invalidation seems like a decent heuristic for that goal. That is, if the user has not touched the set of files referenced by `invalidation_globs` (assuming https://github.com/pantsbuild/pants/pull/20914 has landed) then it won't rerun the relevant `shell_command` / `adhoc_tool`. And this provides a way for us to invalidate the memoized execution in pantsd when the the mtime's do change. mtime invalidation will likely cause more executions to occur than pure content hash based caching, but it should allow a developer who is running Pants repeatedly to avoid invoking the external process every time they run a Pants command.

Reply:
Maybe we should pair on this point?

Reply:
Especially around the interaction of pantsd memoization and cache scope's effect on process caching.